crates/provider/src/layers/cache.rs

```diff
@@ -0,0 +1,753 @@
+        let cache = $cache.clone();
+        ProviderCall::BoxedFuture(Box::pin(async move {
+            let client = client?;
+
+            let result = client.request($req.method(), $req.params()).map_params(|params| {
+                ParamsWithBlock { params, block_id: $req.block_id.unwrap_or(BlockId::latest()) }
+            });
+
+            let res = result.await?;
+            // Insert into cache.
+            let json_str = serde_json::to_string(&res).map_err(TransportErrorKind::custom)?;
+            let hash = $req.params_hash()?;
+            let _ = cache.put(hash, json_str);
+
+            Ok(res)
+        }))
+    }};
+}
+
+/// Attempts to fetch the response from the cache by using the hash of the request params.
+///
+/// Fetches from the RPC in case of a cache miss
+///
+/// This helps overriding [`Provider`] methods that return `RpcWithBlock`.
+macro_rules! cache_rpc_call_with_block {
+    ($cache:expr, $client:expr, $req:expr) => {{
+        if $req.has_block_tag() {
+            return rpc_call_with_block!($cache, $client, $req);
+        }
+
+        let hash = $req.params_hash().ok();
+
+        if let Some(hash) = hash {
+            if let Ok(Some(cached)) = $cache.get_deserialized(&hash) {
+                return ProviderCall::BoxedFuture(Box::pin(async move { Ok(cached) }));
+            }
+        }
+
+        rpc_call_with_block!($cache, $client, $req)
+    }};
+}
+
+#[cfg_attr(target_arch = "wasm32", async_trait::async_trait(?Send))]
+#[cfg_attr(not(target_arch = "wasm32"), async_trait::async_trait)]
+impl<P, T, N> Provider<T, N> for CacheProvider<P, T, N>
+where
+    P: Provider<T, N>,
+    T: Transport + Clone,
+    N: Network,
+{
+    #[inline(always)]
+    fn root(&self) -> &RootProvider<T, N> {
+        self.inner.root()
+    }
+
+    async fn get_block_by_number(
+        &self,
+        number: BlockNumberOrTag,
+        hydrate: bool,
+    ) -> TransportResult<Option<N::BlockResponse>> {
+        let hash = RequestType::new("eth_getBlockByNumber", (number, hydrate));
+
+        cache_get_or_fetch!(self, hash, self.inner.get_block_by_number(number, hydrate))
+    }
+
+    async fn get_block_by_hash(
+        &self,
+        hash: BlockHash,
+        kind: BlockTransactionsKind,
+    ) -> TransportResult<Option<N::BlockResponse>> {
+        let full = match kind {
+            BlockTransactionsKind::Full => true,
+            BlockTransactionsKind::Hashes => false,
+        };
+
+        let req_hash = RequestType::new("eth_getBlockByHash", (hash, full));
+
+        cache_get_or_fetch!(self, req_hash, self.inner.get_block_by_hash(hash, kind))
+    }
+
+    fn get_block_receipts(
+        &self,
+        block: BlockId,
+    ) -> ProviderCall<T, (BlockId,), Option<Vec<N::ReceiptResponse>>> {
+        let req = RequestType::new("eth_getBlockReceipts", (block,));
+
+        let redirect =
+            !matches!(block, BlockId::Hash(_) | BlockId::Number(BlockNumberOrTag::Number(_)));
```

Comment:
the number case is also slightly problematic, but we can worry about this separately